File: framework/languages/dotnet/samples/Bingo/Server/Play/EntrySpot/Handlers/MatchBingoActorHandler.cs
using Systems.Zlink;
using Systems.Zlink.Codecs.Json;
using Zlink.Framework.Contracts.Actors;
using Zlink.Framework.Contracts.Channels;
using Zlink.Framework.Contracts.Configuration;
using Zlink.Framework.Contracts.Handlers;
using Zlink.Framework.Contracts.Spots;
using Zlink.Framework.Contracts.Streams;
using Zlink.Framework.Contracts.Timers;
using Bingo.Server.Play.Actors;
using Bingo.Server.Play.BingoRoomSpots;
using Bingo.Shared.Configuration;
using Bingo.Shared.Contracts;
using Microsoft.Extensions.Logging;

namespace Bingo.Server.Play.EntrySpot.Handlers;

internal sealed class MatchBingoActorHandler(ILogger<MatchBingoActorHandler> logger)
    : IZLinkEntrySpotActorRequestHandler<BingoEntrySpot, PlayerActor, MatchBingoReq, MatchBingoRes>
{
    public async ValueTask<MatchBingoRes> HandleAsync(
        BingoEntrySpot entrySpot,
        PlayerActor actor,
        ZLinkSpotActorRequestContext context,
        MatchBingoReq message,
        CancellationToken cancellationToken)
    {
        _ = context;
        logger.LogInformation("match: actor request. actor={ActorId}, mode={Mode}", actor.ActorId, message.Mode);
        var matched = await entrySpot.Context.Outbound.RequestToChannel(
                SampleNames.ApiChannel,
                new MatchBingoApiReq(actor.ActorId, actor.DisplayName, message.Mode))
            .Timeout(SampleTimings.RequestTimeout)
            .SubmitAsync<MatchBingoApiRes>(cancellationToken)
            ;
        logger.LogInformation("match: room allocated. actor={ActorId}, room={RoomId}", actor.ActorId, matched.RoomId);

        var roomRid = RoutingId.FromHex(matched.RoomId);
        var joined = await actor.Context.JoinSpot(
                roomRid,
                new BingoRoomJoinReq(matched.RoomId, actor.ActorId, actor.DisplayName).Encode())
            .Timeout(SampleTimings.RequestTimeout)
            .SubmitAsync(cancellationToken)
            ;
        logger.LogInformation("match: actor joined room. actor={ActorId}, room={RoomId}", actor.ActorId, matched.RoomId);

        return new MatchBingoRes(matched.RoomId, joined.Reply.Decode<BingoRoomJoinRes>().State);
    }
}

File: framework/languages/node/samples/Bingo.Ts/Server/Play/Infrastructure/ZLink/Spots/EntrySpot/Handlers/match-bingo-actor-handler.ts
import { Inject } from '@nestjs/common';
import {
  ZLINK_ACTOR_MANAGER,
  ZLINK_CHANNEL_CLIENT,
  zlinkEntrySpotActorRequestHandler
} from '@zlink-systems/nestjs';
import { BingoEntrySpot } from '../bingo-entry-spot';
import { PlayerActor } from '../../../Actors/player-actor';
import { PacketNames } from '../../../../../../../Shared/Contracts/messages';
import {
  BingoRoomJoinReq,
  MatchBingoApiReq,
  MatchBingoRes
} from '../../../../../../../Shared/Contracts/bingo-messages.generated';
import { SampleNames, SampleTimings } from '../../../../../../Configuration/sample-names';
import type {
  ZLinkActorManager,
  ZLinkChannelClient,
  ZLinkEntrySpotActorRequestHandler,
  ZLinkSpotActorRequestContext
} from '@zlink-systems/framework';
import type { PlayerActor as PlayerActorType } from '../../../Actors/player-actor';
import type {
  BingoRoomJoinRes,
  MatchBingoApiRes,
  MatchBingoReq
} from '../../../../../../../Shared/Contracts/messages';

@zlinkEntrySpotActorRequestHandler({
  actor: () => PlayerActor,
  entrySpot: () => BingoEntrySpot,
  packetName: PacketNames.matchBingoReq
})
class MatchBingoActorHandler
  implements ZLinkEntrySpotActorRequestHandler<PlayerActorType, MatchBingoReq, MatchBingoRes> {
  constructor(
    @Inject(ZLINK_ACTOR_MANAGER) private readonly actors: ZLinkActorManager,
    @Inject(ZLINK_CHANNEL_CLIENT) private readonly channels: ZLinkChannelClient
  ) {}

  async handle(
    actor: PlayerActorType,
    context: ZLinkSpotActorRequestContext,
    request: MatchBingoReq
  ): Promise<MatchBingoRes> {
    console.error(`bingo-match request actor=${actor.actorId}`);
    const actorRef = await this.actors.find(SampleNames.roomSpotNode, actor.actorId);
    if (actorRef === undefined) {
      throw new Error(`Bingo actor '${actor.actorId}' is not registered.`);
    }
    const matched = await this.channels
      .requestToChannel(
        SampleNames.roomSpotNode,
        SampleNames.apiChannel,
        new MatchBingoApiReq({
          actorId: actor.actorId,
          displayName: actor.displayName,
          actorNodeRid: String(actorRef.nodeRid),
          mode: request.mode
        })
      )
      .timeout(SampleTimings.requestTimeout)
      .submit<MatchBingoApiRes>();
    const joined = await actor.context
      .joinSpot(matched.roomId, new BingoRoomJoinReq({
        roomId: matched.roomId,
        actorId: actor.actorId,
        displayName: actor.displayName,
        observeOnly: false
      }))
      .timeout(SampleTimings.requestTimeout)
      .submit<BingoRoomJoinRes>();
    if (joined.status === 'rejected') {
      throw new Error(`Room ${matched.roomId} rejected actor '${actor.actorId}'.`);
    }
    const response = new MatchBingoRes({
      roomId: matched.roomId,
      state: joined.reply.state,
      roomOwnerNodeRid: matched.roomOwnerNodeRid
    });
    console.error(`bingo-match reply actor=${actor.actorId} room=${response.roomId}`);
    void context;
    return response;
  }
}

export { MatchBingoActorHandler };
